include/wil/filesystem.h

```diff
@@ -1043,8 +1046,224 @@ namespace wil
         THROW_IF_FAILED(GetFileInfoNoThrow<infoClass>(fileHandle, result));
         return result;
     }
+
+    // Helpers to make the CreateFileW API easier to use. This segregates the OPEN_EXISTING "open an existing file"
+    // cases from the "create a new file" that has 4 variations represented in the create_file_create_behavior enum.
+
+    // https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew?devlangs=cpp&f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(FILEAPI%252FCreateFileW)%3Bk(CreateFileW)%3Bk(DevLang-C%252B%252B)%3Bk(TargetOS-Windows)%26rd%3Dtrue
+
+    struct file_and_error_result
+    {
+        file_and_error_result(HANDLE file_handle, DWORD error) : file(file_handle), last_error(error)
+        {
+        }
+
+        wil::unique_hfile file;
+        DWORD last_error{};
+    };
+
+    /** Non-throwing open existing using OPEN_EXISTING.
+    ~~~
+    auto handle = wil::try_open_file(filePath.c_str());
+    ~~~
+    */
+    inline file_and_error_result try_open_file(PCWSTR path, DWORD dwDesiredAccess = FILE_READ_ACCESS,
+        DWORD dwShareMode = FILE_SHARE_READ,
+        bool inheritHandle = false,
+        DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL) noexcept
+    {
+        SECURITY_ATTRIBUTES secAttributes{ sizeof(secAttributes) };
+        secAttributes.bInheritHandle = inheritHandle;
+        return { CreateFileW(path, dwDesiredAccess, dwShareMode, &secAttributes, OPEN_EXISTING, dwFlagsAndAttributes, nullptr),
+            ::GetLastError() };
+    }
+
+    /** open existing using OPEN_EXISTING, throws on error.
+    ~~~
+    auto handle = wil::open_file(filePath.c_str());
+    ~~~
+    */
+    inline wil::unique_hfile open_file(PCWSTR path, DWORD dwDesiredAccess = FILE_READ_ACCESS,
+        DWORD dwShareMode = FILE_SHARE_READ,
+        bool inheritHandle = false,
+        DWORD dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL)
+    {
+        auto result = try_open_file(path, dwDesiredAccess, dwShareMode, inheritHandle, dwFlagsAndAttributes);
+        THROW_WIN32_IF(result.last_error, !result.file.is_valid());

```

Comment:
Probably needs "if exceptions enabled" pre-processor check

Reply:
it seems to compile. I'll leave it I guess.

Reply:
Oh, that's because you defined these functions inside of `#ifdef _CPPUNWIND` (starts on line 1027). Seems like we'd probably want the `try_` versions to be available without exceptions, right?

Reply:
I'm not very sympathetic for projects that build with exceptions disabled. for them, go direct to CreateFileW